Add Qwen2-Audio

faychu opened the pull request:
This PR adds the support of codes for the Qwen2-Audio models. For information about Qwen2-Audio, please visit https://github.com/QwenLM/Qwen2-Audio. @ArthurZucker

faychu committed:
add qwen2audio

faychu committed:
Update check_repo.py

amyeroberts commented:
cc @sanchit-gandhi @kamilakesbi

faychu committed:
fix style

faychu committed:
fix test

sanchit-gandhi commented:
Super cool @faychu! Let us know if you have any questions regarding the model integration - more than happy to help here. Otherwise, ping us when this one is ready for review and we can take a look!

Also cc @ylacombe @Vaibhavs10 for viz

faychu committed:
fix style

faychu committed:
add model size

ylacombe commented:
Hey @faychu, thanks for adding the model directly into transformers!

If your testing the model locally, you need to follow the instruction (`huggingface-cli login`) in order to be able to load the weights from the hub. Don't hesitate to read more about it [here](https://huggingface.co/docs/huggingface_hub/en/guides/cli#huggingface-cli-login) or ping us again in this PR.

If you're talking about the online CI tests, I'm not aware of being able to resolve this issue without making the model public (maybe @amyeroberts or @ydshieh have a solution?). 
But this shouldn't be a blocker for you if you can test everything locally. We can review the PR without having a fully green CI online, and in the meantime you can run slow tests locally, as long as you have a GPU.

Once the PR is approved, and every slow tests pass locally on your side, we can make the model public at the last moment just before integrating the PR, what do you think?

faychu commented:
Hey @ylacombe 

Great! Thank you. I have successfully passed all the tests locally following the instructions on https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request, and I believe the code is ready.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+            if output_attentions:
+                all_attentions = all_attentions + (layer_outputs[1],)
+
+        hidden_states = hidden_states.permute(0, 2, 1)
+        hidden_states = self.avg_pooler(hidden_states)
+        hidden_states = hidden_states.permute(0, 2, 1)
```
Besides this particular snippet, this corresponds almost entirely to Whisper implementation, right?

Small questions on my side, are you finetuning the Whisper weights? If so, I wonder if it'd make sense to use `WhisperEncoder` directly, since it'd make the implementation much more easier to read and to use!

I think it would make sense to either:
* Add a config parameter to WhisperConfig which allows [to not use a `layer_norm` in the Encoder](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/src/transformers/models/whisper/modeling_whisper.py#L1078), and thus to rewrite the `Qwen2AudioEncoderModel` with 3 objects: 1. an `audio_encoder`, i.e an AutoModel that loads WhisperEncoder. 2. `avg_pooler` 3. `layer_norm`.
* Or a more straightforward method: add the possibility to add an `avg_pooler` [here](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/src/transformers/models/whisper/modeling_whisper.py#L1078) to the `WhisperEncoder` code, and use it instead of `Qwen2AudioEncoderModel`.

faychu replied:
Thanks for your suggestions. Yes, I finetune the Whisper weights. Enable whisper encoder to accept any chunk length

faychu replied:
Thanks for your suggestions.

faychu replied:
We sincerely appreciate your suggestion! In the future, we may update the audio encoder to support arbitrary-length chunks and potentially include more types of position embeddings. To avoid any possible impact on the Whisper model branch, we would prefer to keep `Qwen2AudioEncoder`.

ylacombe replied:
Thanks for the explanation! 

Given this new information, I think it would make sense to keep the code as it is. However, I'd still be in favor of making it clearer that it's mostly the Whisper encoder model.

To do so, you can add "# Copied from..." and "# Ignore copy" strategically:
1. in the init and in the forward
2. And add "# Ignore copy" where relevant, i.e where using `avg_pooler`


Also cc @amyeroberts and @sanchit-gandhi for a second opinion on this thread

amyeroberts replied:
I'd be pro this suggestion: # Copied from..." and "# Ignore copy" strategically.

faychu replied:
Thank you, I will add these as you suggested.

faychu replied:
already added `# Copied from ...` and `# Ignore copy` here.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+    @property
+    def _supports_sdpa(self):
+        """
+        Retrieve language_model's attribute to check whether the model supports
```
Great!

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+        return self.language_model._supports_sdpa
+
+
+QWEN2AUDIO_INPUTS_DOCSTRING = r"""
```
It's really clean! I'd simply switch the `attention_mask` and the `feature_attention_mask` here and in the method, in order to keep the order text->audio

faychu replied:
Thanks. I've switched them.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+        self._padding_side = "left"  # set it to left by default, user can use setter to change padding_sides
+        self.post_init()
+
+    @property
+    def padding_side(self):
+        return self._padding_side
+
+    @padding_side.setter
+    def padding_side(self, padding_side: str):
+        if padding_side not in ["left", "right"]:
+            raise ValueError(f"{padding_side} is not `left` or `right`.")
+        self._padding_side = padding_side
+
+    def get_input_embeddings(self):
+        return self.language_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.language_model.set_input_embeddings(value)
+
+    def get_output_embeddings(self):
+        return self.language_model.get_output_embeddings()
+
+    def set_output_embeddings(self, new_embeddings):
+        self.language_model.set_output_embeddings(new_embeddings)
+
+    def set_decoder(self, decoder):
+        self.language_model.set_decoder(decoder)
+
+    def get_decoder(self):
+        return self.language_model.get_decoder()
+
+    def tie_weights(self):
+        return self.language_model.tie_weights()
+
+    def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_multiple_of=None) -> nn.Embedding:
+        model_embeds = self.language_model.resize_token_embeddings(new_num_tokens, pad_to_multiple_of)
+        # update vocab size
+        self.config.text_config.vocab_size = model_embeds.num_embeddings
+        self.vocab_size = model_embeds.num_embeddings
```
Let's add `Copied` statements here as well (same)
```suggestion
    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.get_input_embeddings
    def get_input_embeddings(self):
        return self.language_model.get_input_embeddings()

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.set_input_embeddings
    def set_input_embeddings(self, value):
        self.language_model.set_input_embeddings(value)

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.get_output_embeddings
    def get_output_embeddings(self):
        return self.language_model.get_output_embeddings()

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.set_output_embeddings
    def set_output_embeddings(self, new_embeddings):
        self.language_model.set_output_embeddings(new_embeddings)

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.set_decoder
    def set_decoder(self, decoder):
        self.language_model.set_decoder(decoder)

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.get_decoder
    def get_decoder(self):
        return self.language_model.get_decoder()

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.tie_weights
    def tie_weights(self):
        return self.language_model.tie_weights()

    # Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration.resize_token_embeddings
    def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_multiple_of=None) -> nn.Embedding:
        model_embeds = self.language_model.resize_token_embeddings(new_num_tokens, pad_to_multiple_of)
        # update vocab size
        self.config.text_config.vocab_size = model_embeds.num_embeddings
        self.vocab_size = model_embeds.num_embeddings
```

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+            state=attention_mask,
+        )
+
+    def prepare_inputs_for_generation(
```
Let's add a Copied statement here since it looks really similar to [here](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/src/transformers/models/llava_next/modeling_llava_next.py), if I'm not mistaken

faychu replied:
Already add a Copied statement here [add copied from to prepare_inputs_for_generation](https://github.com/huggingface/transformers/pull/32137/commits/22db9401b9d0d1ad9bd65d8311eae836274adafb)

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+            else:
+                cache_length = past_length = past_key_values[0][0].shape[2]
+
+            if input_features is not None and kwargs.get("state") is not None:
+                attention_mask = kwargs["state"]
+                attention_mask = torch.cat(
+                    [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
+                )
```
It's quite a difference, that could justify not using the Copied statements, but since we're seeing huge changes in the generation logic, I think we'd rather have "# Ignore copy" statements, that will facilitate future incorporation of changes.


WDYT @gante ?

gante replied:
Yes, let's do that, and let's also add a comment explaining the differences for our future selves (`feature_attention_mask` is piped around) 🤗

faychu replied:
@gante  @ylacombe  Thanks, aready added a Copied statement here [add copied from to prepare_inputs_for_generation](https://github.com/huggingface/transformers/pull/32137/commits/22db9401b9d0d1ad9bd65d8311eae836274adafb), and also added a comment as you suggest.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+                audio_outputs = self.audio_tower(
+                    input_features, attention_mask=audio_attention_mask
+                )  # [3, 1500, 1280]
+                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
```
You don't have to pass output_hidden_states=True to get `last_hidden_state` though!

faychu replied:
Already fixed it.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+                )  # [3, 1500, 1280]
+                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
+                selected_audio_feature = audio_outputs.last_hidden_state
+                audio_features = self.multi_modal_projector(selected_audio_feature)  # [3, 750, 4096]
```
(nit)
You can add the expected shape (batch_size, sequence_length, hidden_dim) if you want, but let's remove numbers!

```suggestion
                audio_features = self.multi_modal_projector(selected_audio_feature) 
```

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+
+                audio_outputs = self.audio_tower(
+                    input_features, attention_mask=audio_attention_mask
+                )  # [3, 1500, 1280]
```
Same here (nit)
```suggestion
                )
```

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+        ) < num_audio_tokens.unsqueeze(1)
+        masked_audio_features = audio_features[audio_features_mask].view(-1, embed_dim)
+        batch_size, sequence_length = input_ids.shape
+        # left_padding = not torch.sum(input_ids[:, -1] == torch.tensor(self.pad_token_id))
```
(nit)
let's remove this?
```suggestion
```

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+        special_audio_token_mask = input_ids == self.config.audio_token_index
+        num_special_audio_tokens = torch.sum(special_audio_token_mask, dim=-1)
+
+        # In case the Audio model or the Language model has been offloaded to CPU, we need to manually
```
Do we support offloading?

faychu replied:
I think the inputs may be offloaded to CPU.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+        self.vocab_size = model_embeds.num_embeddings
+        return model_embeds
+
+    def _merge_input_ids_with_audio_features(
```
This is already well commented, but let's add a docstrings as well, to make it clearer please!

faychu replied:
Thanks, already added it [here](https://github.com/huggingface/transformers/pull/32137/commits/07e4c0f07d85da064d469cdda17f1acf35231b6b).

ylacombe commented on docs/source/en/index.md:
```diff
@@ -254,6 +254,8 @@ Flax), PyTorch, and/or TensorFlow.
 |                       [QDQBert](model_doc/qdqbert)                       |       ✅        |         ❌         |      ❌      |
 |                         [Qwen2](model_doc/qwen2)                         |       ✅        |         ❌         |      ❌      |
+|                   [Qwen2Audio](model_doc/qwen2_audio)                    |       ✅        |         ❌         |      ❌      |
+|            [Qwen2AudioEncoder](model_doc/qwen2_audio_encoder)            |       ✅        |         ❌         |      ❌      |
```
I don't think we should expose Qwen2AudioEncoder as it is a part of the model, and not usable by itself. Let's not expose it

ylacombe replied:
I think you can add it to [`PRIVATE_MODELS`](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/utils/check_repo.py#L62), but I'd like to double-check with @amyeroberts.

Another solution would be to remove `qwen2_audio_encoder` from the `configuration_auto` and `modeling_auto` classes, but then you wouldn't be able to load the config as you do

amyeroberts replied:
Yes, adding to `PRIVATE_MODELS` is the way to go if you want to have a separate config. We do this for e.g. CLIP's text model so OK imo

faychu replied:
Thanks, already fixed it [here](https://github.com/huggingface/transformers/pull/32137/commits/1a86c116737f9c778dba77c4a6136c40170f7786).

ylacombe commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,79 @@
+
+## Overview
+
+The Qwen2Audio is the new model series of large audio-language models from the Qwen team.
```
Let's not forget to add a reference to the technical paper, as well as the abstract. For example, refer to `llama.md`

ylacombe commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,79 @@
+
+## Usage tips
+
+`Qwen2-Audio-7B` and `Qwen2-Audio-7B-Instruct` can be found on the [Huggingface Hub](https://huggingface.co/Qwen)
```
This is a great first start, but the model docs would benefit from having more details on how to use the model in the two interactions modes that are highlighted in the paper.

You could also add some proper usage tips and resources, as well as a description of the architecture. See for example `seamless_m4t_v2.md`. What do you think?

Note that you can then almost copy / paste what you've done here to the model card of the weights, once they'll be public!

faychu replied:
Already [add more details to docs](https://github.com/huggingface/transformers/pull/32137/commits/d3e5e716426e07c655d37ac636fe27639f39186d).

ylacombe reviewed (commented):
Hey @faychu, thanks for opening such a clean PR! 
This is a really impressive PR, you made great use of other modeling codes!

I've left a few comments, but I think we should first decide whether we want to keep the `Qwen2AudioEncoder` as it is,  modify it or remove it whatsoever! Let continue the discussion directly in the concerned comments, but happy to hear your thoughts on this.

faychu committed:
Qwen2AudioEncoderModel->Qwen2AudioEncoder; add copy info

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
switch the attention_mask and the feature_attention_mask

faychu commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1239 @@
+    state: Optional[torch.FloatTensor] = None
+
+
+class Qwen2AudioAttention(nn.Module):
```
I've added `# Copied from ...` as you suggested.

faychu committed:
add to PRIVATE_MODELS in check_repo.py; add to MODEL_NAMES_TO_IGNORE in check_table.py

faychu committed:
fix initialization

faychu committed:
update chat_template

faychu committed:
fix consistency issue after copy

faychu committed:
add docstrings to _merge_input_ids_with_audio_features

faychu committed:
add copied from to prepare_inputs_for_generation

faychu committed:
add more details to docs

faychu commented:
Hey @ylacombe , I've made the changes you suggested to the code. Could you please take another look to see if it resolves the issues mentioned?

faychu committed:
rm comment

faychu committed:
add init_std

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32137). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ylacombe commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+inputs = processor(text=text, audios=audios, return_tensors="pt", padding=True)
+inputs.input_ids = inputs.input_ids.to("cuda")
+
+generate_ids = model.generate(**inputs, max_length=256)
+generate_ids = generate_ids[:, inputs.input_ids.size(1):]
```
As discussed in the modeling test, `generate_ids` should be made of entirely newly generated tokens, and you shouldn't have to pass the input ids length to get the generated tokens.

ylacombe replied:
Thanks, I haven't caught that! 

cc @amyeroberts and @ArthurZucker , is this expected? Thanks for your insights

gante replied:
This has always been the behavior in `generate` for decoder-only models -- the model continues the input, so it makes sense for the whole thing to stay on a single tensor. 

This is different from the case with encoder-decoder models, where the generated content starts anew (because the input is NOT fed to the decoder). If you pass a started decoder sequence to an encoder-decoder model through `decoder_input_ids`, you'll also see this sequence in the output 🤗 

Because it is simple to crop and `generate` is meant to be a lower-level API, I'd be biased against adding a flag to output exclusively new tokens, unless there is a big demand for this functionality.

ylacombe commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1326 @@
+            feature_attention_mask = feature_attention_mask.to(target_device)
+
+        if inputs_embeds is None:
+            # 1. Extra the input embeddings
```
(nit)
```suggestion
            # 1. Extract the input embeddings
```

ylacombe reviewed (commented):
Hi @faychu, 
thanks for iterating, this looks really good! I appreciate the efforts made on the .md!

I've left a few comments, but I'm also left with a big question that appears throughout the review:

Why does the output of `generate` contains input ids ? 

Do you have any intuition as to where and why it appears? If you do, let's try to solve this before asking for a core maintainer review!

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu commented on src/transformers/models/qwen2_audio/processing_qwen2_audio.py:
```diff
@@ -0,0 +1,158 @@
+        ]
+
+        result_with_id = template.render(messages=messages, add_generation_prompt=True, add_audio_id=True)
+        result_without_id = template.render(messages=messages, add_generation_prompt=True, add_audio_id=False)
```
Actually, the add_audio_id is always true in qwen2audio. Should we remove the parameter?

ylacombe replied:
I think we should indeed!

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
Update src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

Co-authored-by: Yoach Lacombe <52246514+ylacombe@users.noreply.github.com>

faychu committed:
update

ylacombe commented:
Hey @amyeroberts or @ArthurZucker, thanks to @faychu's great work, I think the PR is ready for your review!

BTW, there's still the pending question about input ids being part of the generated ids (cf [here](https://github.com/huggingface/transformers/pull/32137#discussion_r1695247851)) that needs your insights !

amyeroberts commented:
@ylacombe That's great news! 

Regarding final review from a core maintainer: 
* If you think it's mergeable then it helps us to know this with a final review and approval from someone else 
* All tests should be passing. I triggered a re-run of the CI tests which looked like they were transient. At the moment, there's still a model-specific test which is failing `tests/models/qwen2_audio/test_modeling_qwen2_audio.py::Qwen2AudioForConditionalGenerationModelTest::test_batching_equivalence` 

For the input_ids in the outputs, this is unfortunately inconsistent across models in the library cc @gante here - should we be aiming to just return the generated ids?

ylacombe commented:
Hey @amyeroberts, thanks for your insights!

In that case, let's wait a bit for your review, I'll provide a final review when the current failing test is green again, and when the generated ids issue is resolved.

> For the input_ids in the outputs, this is unfortunately inconsistent across models in the library cc @gante here - should we be aiming to just return the generated ids?

This is what I thought as well, but what about @faychu's [rightful observation](https://github.com/huggingface/transformers/pull/32137#discussion_r1695391004) that Llava also includes the input ids ?

amyeroberts commented:
> This is what I thought as well, but what about @faychu's https://github.com/huggingface/transformers/pull/32137#discussion_r1695391004 that Llava also includes the input ids ?

Sorry, I wasn't clear. Yes, @faychu is right, some models do return the input_ids in the generated sequence, some don't. This has been a problem for building the mulitmodal pipelines involving VLMs. To the best of my knowledge, we don't normally return the input input_ids, but I'm not 100% sure tbh so was pinging @gante as he's more likely to know :)

ylacombe reviewed (approved):
Hey @amyeroberts,
Looks like the input ids "issue" was the expected behaviour.

The remaining failing tests are due to the weights repository being private and pass in the contributor's local tests.

Could you review when you have time? @faychu will have to make the model public a bit before merging so that everything's green!

amyeroberts commented on tests/models/qwen2_audio/test_modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,310 @@
+                for ele in message["content"]:
+                    if ele["type"] == "audio":
+                        audios.append(
+                            ffmpeg_read(
```
Do we have any other way to read some audio for these tests @ylacombe? e.g. through a dataset

I don't think this is a good pattern for three reasons:
* Importing a utils from the pipelines is a sign that our code isn't well contained -- a model sits lower in abstraction that pipelines and shouldn't need to know about them at all. The utility should probably be moved 
* This function uses subprocesses - anything which forces us out of python land is a code smell
* This assumes ffmpeg is available in the environment. It should be protected with a `@require_ffmpeg` decorator

ylacombe replied:
We actually do, cf how it's done for Whisper: https://github.com/huggingface/transformers/blob/9451a385261b30e7319a2c93285ab76161e8c003/tests/models/whisper/test_modeling_whisper.py#L1861-L1866

ylacombe replied:
That said, the reasons you gave against this pattern make total sense. I agree we should remove this pattern, at least in the modeling test.

In terms of how to actually use the model - as shown in the `.md` and discussed [here](https://github.com/huggingface/transformers/pull/32137#discussion_r1699130692) - I actually think `ffmpeg_read` is a great method because it performs a few operation under the hood: reading from whatever source there is, resampling the waveform and outputting a fp32 numpy array.

Since `ffmpeg` is quite a common utility to have, I feel like the strongest argument against it is that it uses sub-processes.

 Could we show in the `.md` how to use the model:
1. without using `ffmpeg_read`: i.e we could use [scipy read method](https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.wavfile.read.html#scipy.io.wavfile.read) or `torchaudio.load` and show how to resample the audio
2. using `ffmpeg_read`: to show how to use the model from a list of urls

WDYT @amyeroberts ?

faychu replied:
Hey, @ylacombe @amyeroberts I've removed `ffmpeg_read` and replaced it with `liborsa` to read the audio. Can you help me take a look again? see this [commit](https://github.com/huggingface/transformers/pull/32137/commits/8d029b808f9e7f23701782f26526f7f6208fd7ca)

amyeroberts commented on tests/models/qwen2_audio/test_modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,310 @@
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
+
+    @unittest.skip(reason="Compile not yet supported because in LLava models")
+    def test_sdpa_can_compile_dynamic(self):
+        pass
+
+    @unittest.skip(reason="Compile not yet supported because in LLava models")
+    def test_sdpa_can_dispatch_on_flash(self):
+        pass
```
To update? This isn't a llava model

Were these tests just copy-pasted? Have you tried running without the skips?

faychu replied:
I've tried running without the skips but failed, so I update the name LLava -> Qwen2Audio

amyeroberts commented on tests/models/qwen2_audio/test_modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,310 @@
+        self.model_tester = Qwen2AudioModelTester(self)
+        self.config_tester = ConfigTester(self, config_class=Qwen2AudioConfig, has_text_modality=False)
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
```
Have these been tested without the skips?

faychu replied:
Yes, I've remove the skips.

amyeroberts commented on src/transformers/models/qwen2_audio/processing_qwen2_audio.py:
```diff
@@ -0,0 +1,154 @@
+        audios = kwargs.pop("audios", None)
+        sampling_rate = kwargs.pop("sampling_rate", None)
+        text = kwargs.pop("text", None)
+        if len(args) > 0:
+            audios = args[0]
+            args = args[1:]
```
this is super dirty. It is far better to have explicit keyword arguments in the call signature

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+
+The Qwen2-Audio is the new model series of large audio-language models from the Qwen team. Qwen2-Audio is capable of accepting various audio signal inputs and performing audio analysis or direct textual responses with regard to speech instructions. We introduce two distinct audio interaction modes:
+
+* voice chat: users can freely engage in voice interactions with Qwen2-Audio without text input;
+
+* audio analysis: users could provide audio and text instructions for analysis during the interaction;
```
nit 

```suggestion
* voice chat: users can freely engage in voice interactions with Qwen2-Audio without text input
* audio analysis: users could provide audio and text instructions for analysis during the interaction
```

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+    if isinstance(message["content"], list):
+        for ele in message["content"]:
+            if ele["type"] == "audio":
+                audios.append(ffmpeg_read(requests.get(ele['audio_url']).content, sampling_rate=processor.feature_extractor.sampling_rate))
```
```suggestion
                audios.append(
                    ffmpeg_read(requests.get(ele['audio_url']).content,
                    sampling_rate=processor.feature_extractor.sampling_rate)
                )
```

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+            if ele["type"] == "audio":
+                audios.append(ffmpeg_read(requests.get(ele['audio_url']).content, sampling_rate=processor.feature_extractor.sampling_rate))
+
+inputs = processor(text=text, audios=audios, return_tensors="pt", padding=True)
```
```suggestion
inputs = processor(text=text, audio=audios, return_tensors="pt", padding=True)
```

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
````diff
@@ -0,0 +1,188 @@
+In the audio analysis, users could provide both audio and text instructions for analysis:
+```python
+from transformers import Qwen2AudioForConditionalGeneration, AutoProcessor
+from transformers.pipelines.audio_utils import ffmpeg_read
````
Same here - we shouldn't be importing utilities from pipeline - this is an indication the logic is in the wrong place

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+    if isinstance(message["content"], list):
+        for ele in message["content"]:
+            if ele["type"] == "audio":
+                audios.append(ffmpeg_read(requests.get(ele['audio_url']).content, sampling_rate=processor.feature_extractor.sampling_rate))
```
line is too long  - should be split across lines

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,188 @@
+        if isinstance(message["content"], list):
+            for ele in message["content"]:
+                if ele["type"] == "audio":
+                    audios.append(ffmpeg_read(requests.get(ele['audio_url']).content, sampling_rate=processor.feature_extractor.sampling_rate))
```
same here

amyeroberts commented on src/transformers/models/qwen2_audio/configuration_qwen2_audio.py:
```diff
@@ -0,0 +1,203 @@
+            The config object or dictionary of the audio backbone.
+        text_config (`Union[AutoConfig, dict]`, *optional*, defaults to `LlamaConfig`):
+            The config object or dictionary of the text backbone.
+        ignore_index (`int`, *optional*, defaults to -100):
+            The ignore index for the loss function.
```
Is this necessary to add? Is it often different for these models? It's normally -100 by default

amyeroberts commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1324 @@
+
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+        state (`torch.FloatTensor`, *optional*):
```
Can we use a better name here? `state` isn't very descriptive `attention_mask` would be better

faychu replied:
@amyeroberts  Here we use `state` as the parameter name because we need to update the attention mask during the generate process, and we need to pass the updated attention mask back. The `generate` function only has the `state` interface for passing it.  see [_update_model_kwargs_for_generation](https://github.com/faychu/transformers/blob/qwen2-audio/src/transformers/generation/utils.py#L674)

amyeroberts replied:
In this case I would say that the model should then override `_update_model_kwargs_for_generation` to handle `attention_mask` and then call the parents' method as necessary. `state` is used for stateful models like mamba and RWKV

amyeroberts commented on src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,1324 @@
+        self.vocab_size = model_embeds.num_embeddings
+        return model_embeds
+
+    def _merge_input_ids_with_audio_features(
```
What are the differences here with the similar method in llava?

faychu replied:
@amyeroberts the shape of audio feature is different with image, and we support variable length audio tokens.

amyeroberts reviewed (commented):
Thanks for all the work adding this model! Overall looks great! 

Main comment is about the ffmpeg utility

faychu committed:
Update docs/source/en/model_doc/qwen2_audio.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faychu committed:
update tests

faychu committed:
rm ignore_index

faychu committed:
update processor

faychu committed:
rm ffmpeg_read

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,201 @@
+model = Qwen2AudioForConditionalGeneration.from_pretrained("Qwen/Qwen2-Audio-7B-Instruct", device_map="auto")
+
+conversation = [
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/guess_age_gender.wav"},
+            ]},
+            {"role": "assistant", "content": "Yes, the speaker is female and in her twenties."},
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/translate_to_chinese.wav"},
+            ]},
+        ]
```
```suggestion
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/guess_age_gender.wav"},
    ]},
    {"role": "assistant", "content": "Yes, the speaker is female and in her twenties."},
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/translate_to_chinese.wav"},
    ]},
]
```

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,201 @@
+processor = AutoProcessor.from_pretrained("Qwen/Qwen2-Audio-7B-Instruct")
+model = Qwen2AudioForConditionalGeneration.from_pretrained("Qwen/Qwen2-Audio-7B-Instruct", device_map="auto")
+
+conversation = [
+            {'role': 'system', 'content': 'You are a helpful assistant.'}, 
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/glass-breaking-151256.mp3"},
+                {"type": "text", "text": "What's that sound?"},
+            ]},
+            {"role": "assistant", "content": "It is the sound of glass shattering."},
+            {"role": "user", "content": [
+                {"type": "text", "text": "What can you do when you hear that?"},
+            ]},
+            {"role": "assistant", "content": "Stay alert and cautious, and check if anyone is hurt or if there is any damage to property."},
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/1272-128104-0000.flac"},
+                {"type": "text", "text": "What does the person say?"},
+            ]},
+        ]
```
```suggestion
conversation = [
    {'role': 'system', 'content': 'You are a helpful assistant.'}, 
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/glass-breaking-151256.mp3"},
        {"type": "text", "text": "What's that sound?"},
    ]},
    {"role": "assistant", "content": "It is the sound of glass shattering."},
    {"role": "user", "content": [
        {"type": "text", "text": "What can you do when you hear that?"},
    ]},
    {"role": "assistant", "content": "Stay alert and cautious, and check if anyone is hurt or if there is any damage to property."},
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/1272-128104-0000.flac"},
        {"type": "text", "text": "What does the person say?"},
    ]},
]
```

amyeroberts commented on docs/source/en/model_doc/qwen2_audio.md:
```diff
@@ -0,0 +1,201 @@
+processor = AutoProcessor.from_pretrained("Qwen/Qwen2-Audio-7B-Instruct")
+model = Qwen2AudioForConditionalGeneration.from_pretrained("Qwen/Qwen2-Audio-7B-Instruct", device_map="auto")
+
+conversation1 = [
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/glass-breaking-151256.mp3"},
+                {"type": "text", "text": "What's that sound?"},
+            ]},
+            {"role": "assistant", "content": "It is the sound of glass shattering."},
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/f2641_0_throatclearing.wav"},
+                {"type": "text", "text": "What can you hear?"},
+            ]}
+        ]
+
+conversation2 = [
+            {"role": "user", "content": [
+                {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/1272-128104-0000.flac"},
+                {"type": "text", "text": "What does the person say?"},
+            ]},
+        ]
```
```suggestion
conversation1 = [
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/glass-breaking-151256.mp3"},
        {"type": "text", "text": "What's that sound?"},
    ]},
    {"role": "assistant", "content": "It is the sound of glass shattering."},
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/f2641_0_throatclearing.wav"},
        {"type": "text", "text": "What can you hear?"},
    ]}
]

conversation2 = [
    {"role": "user", "content": [
        {"type": "audio", "audio_url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen2-Audio/audio/1272-128104-0000.flac"},
        {"type": "text", "text": "What does the person say?"},
    ]},
]
```

amyeroberts commented on src/transformers/models/qwen2_audio/processing_qwen2_audio.py:
```diff
@@ -0,0 +1,177 @@
+        """
+
+        if audios is None and text is None:
+            raise ValueError("You need to specify either an `audio` or `text` input to process.")
+        inputs = self.tokenizer(text, padding=padding, **kwargs)
```
These two lines are not consistent: the error says you can specify either audio or text, but the following line assumes text is not None

amyeroberts commented on tests/models/qwen2_audio/test_modeling_qwen2_audio.py:
```diff
@@ -0,0 +1,301 @@
+            "Generate the caption in English: Glass is breaking.",
+            "Classify the human vocal sound to VocalSound in English: Throat clearing",
+        ]
+        print(self.processor.batch_decode(output, skip_special_tokens=True))
```
```suggestion
```

amyeroberts reviewed (commented):
Looking good! 

Two main things left to address: 
* processor behaviour, assuming `text` inputs 
* `state` var name in model outputs

Before merge, we'll need to update the checkpoints to point to public checkpoints, and run the slow tests (add `run-slow` label to the PR and push a commit with the message `[run_slow] qwen2_audio`)

faychu committed:
Update tests/models/qwen2_audio/test_modeling_qwen2_audio.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faychu committed:
Update docs/source/en/model_doc/qwen2_audio.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faychu committed:
Update docs/source/en/model_doc/qwen2_audio.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faychu committed:
Update docs/source/en/model_doc/qwen2_audio.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

faychu committed:
update

faychu committed:
typo

faychu committed:
[run_slow] qwen2_audio

faychu committed:
[run_slow] qwen2_audio

faychu committed:
[run_slow] qwen2_audio

faychu committed:
fix quality

faychu committed:
[run_slow] qwen2_audio

faychu commented:
@amyeroberts I've addressed all your comments, and pushed a commit with the message `[run_slow] qwen2_audio`, and made all the tests green. But I don't know how to add the `run-slow` label. Have we met the requirements for merging?

faychu committed:
[run_slow] qwen2_audio

amyeroberts commented:
@faychu Great - thanks for pushing that commit! At the moment - some of the multi-gpu tests are failing for the model, so these will need to be resolved before we can merge.

faychu committed:
[run_slow] qwen2_audio

faychu commented:
Hey @amyeroberts, I've resolved the multi-gpu tests locally, and pushed a commit with `[run_slow] qwen2_audio`, but it seems that the multi-gpu tests have not been triggered. What should I do now?

faychu commented:
@ylacombe @amyeroberts It looks like the single-gpu test has been stuck for 15 hours and we need your help.

faychu committed:
add official model

amyeroberts commented:
@faychu Hmmmm - I'm not really sure what caused this. Sometimes there aren't enough machines available for the runs and they get stuck even if you try to trigger a re-run from the workflow page. It's more important that the tests pass for the multi-gpu runs, so if they're OK then I'm pretty sure the single GPU ones will be. 

@ylacombe In this case, are you happy to merge and be the one to follow up on tests if we find there's failures for the single GPU runs?

ylacombe commented:
Merging now!
